index.js

```diff
@@ -501,17 +506,29 @@ module.exports = class Protomux {
     const end = state.end
     let remoteId = c.uint.decode(state)
 
+    let waiting = null
+
     while (state.end > state.start) {
       const len = c.uint.decode(state)
       if (len === 0) {
         remoteId = c.uint.decode(state)
         continue
       }
       state.end = state.start + len
-      this._decode(remoteId, state)
+      // if batch contains more than one message, cork it so we reply back with a batch
+      if (end !== state.end && waiting === null) {
+        waiting = []
+        this.cork()
+      }
+      const p = this._decode(remoteId, state)
+      if (waiting !== null && p !== null) waiting.push(p)
       state.start = state.end
       state.end = end
     }
+
+    if (waiting !== null) {
+      Promise.all(waiting).then(this._uncorkBound, this._safeDestroyBound)
```

Comment:
Note: unsure if relevant/correct, because I don't know this code very well.

If one of the promises in waiting rejects, then the batch is never processed (because `uncork(...)` will never trigger the `--this.corked === 0` case).

I'm not sure if those promises can ever reject, but if they can, that's a big behaviour change. If in a batch of 100 messages, one message fails, before we would still process the other 99, whereas now I think we won't process any.

Reply:
I also think that promises in `waiting` might lack a rejection handler (but that's only relevant if they can reject)

Reply:
Good comment, they cant reject, let me add a comment to clarify